pkg/ingester/ingester_ingest_storage_test.go

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ingester
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/dskit/test"
+	"github.com/grafana/dskit/user"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/twmb/franz-go/pkg/kfake"
+	"github.com/twmb/franz-go/pkg/kmsg"
+	"go.uber.org/atomic"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/querier/api"
+	"github.com/grafana/mimir/pkg/storage/ingest"
+	util_test "github.com/grafana/mimir/pkg/util/test"
+	"github.com/grafana/mimir/pkg/util/testkafka"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+func TestIngester_QueryStream_IngestStorageReadConsistency(t *testing.T) {
+	const (
+		metricName = "series_1"
+	)
+
+	for _, readConsistency := range []string{api.ReadConsistencyEventual, api.ReadConsistencyStrong} {
+		t.Run(fmt.Sprintf("read consistency = %s", readConsistency), func(t *testing.T) {
+			var (
+				cfg     = defaultIngesterTestConfig(t)
+				limits  = defaultLimitsTestConfig()
+				reg     = prometheus.NewRegistry()
+				ctx     = context.Background()
+				series1 = mimirpb.PreallocTimeseries{
+					TimeSeries: &mimirpb.TimeSeries{
+						Labels:  mimirpb.FromLabelsToLabelAdapters(labels.FromStrings(labels.MetricName, metricName)),
+						Samples: []mimirpb.Sample{{TimestampMs: 1000, Value: 10}},
+					},
+				}
+			)
+
+			limits.IngestStorageReadConsistency = readConsistency
+
+			// Create the ingester.
+			overrides, err := validation.NewOverrides(limits, nil)
+			require.NoError(t, err)
+			ingester, kafkaCluster := createTestIngesterWithIngestStorage(t, &cfg, overrides, reg)
+
+			// Mock the Kafka cluster to fail the Fetch operation until we unblock it later in the test.
+			// If read consistency is "eventual" then these failures shouldn't affect queries, but if it's set
+			// to "strong" then a query shouldn't succeed until the Fetch requests succeed.
+			failFetch := atomic.NewBool(true)
+
+			kafkaCluster.ControlKey(int16(kmsg.Fetch), func(kreq kmsg.Request) (kmsg.Response, error, bool) {
+				kafkaCluster.KeepControl()
+
+				if failFetch.Load() {
+					return nil, errors.New("mocked error"), true
+				}
+
+				return nil, nil, false
+			})
+
+			// Start the ingester (after the Kafka cluster has been mocked).
+			require.NoError(t, services.StartAndAwaitRunning(ctx, ingester))
+			defer t.Cleanup(func() {
```

Comment:
do we need both the defer and the Cleanup? (same applies to the Cleanup below)

Reply:
Oh damn... my bad. Removing the defer.